File: srcs/django-files/backend/tpong/game/engine/Vec2.py
import math

class Vec2():
	def __init__(self, x=0, y=0) -> None:
		self.x, self.y = x, y
	
	def __sub__(self, other):
		if isinstance(other, int) or isinstance(other, float):
			return Vec2(self.x - other, self.y - other)
		else:
			return Vec2(self.x - other.x, self.y - other.y)

	def __add__(self, other):
		if isinstance(other, int) or isinstance(other, float):
			return Vec2(self.x + other, self.y + other)
		else:
			return Vec2(self.x + other.x, self.y + other.y)

	def __mul__(self, other):
		if isinstance(other, int) or isinstance(other, float):
			return Vec2(self.x * other, self.y * other)
		else:
			return Vec2(self.x * other.x, self.y * other.y)

	def __truediv__(self, other):
		if isinstance(other, int) or isinstance(other, float):
			return Vec2(self.x / other, self.y / other)
		else:
			return Vec2(self.x / other.x, self.y / other.y)

	def __abs__(self):
		return math.sqrt(self.x**2 + self.y**2)
	
	def __str__(self) -> str:
		return f"{self.x},{self.y}"

	def fromOtherInstance(self, instance):
		self.x = instance.x
		self.y = instance.y

	def copy(self):
		return Vec2(self.x, self.y)

	def distTo(self, other):
		return abs(self - other)
	
	def length(self):
		return abs(self)

	def clamp(self, high, low):
		self.x = max(min(self.x, high), low)
		self.y = max(min(self.y, high), low)

File: srcs/django-files/backend/tpong/static/tpong/vec2.js

export class vec2
{
	constructor(x, y)
	{
		if (x == undefined)
			this.x = 0;
		else
			this.x = x;
		if (y == undefined && x != undefined)
			this.y = x;
		else if (y != undefined)
			this.y = y;
		else
			this.y = 0;
	}
	
	copy()
	{
		return new vec2(this.x, this.y);
	}

	multiply(vec)
	{
		return new vec2(this.x * vec.x, this.y * vec.y);
	}

	divide(vec)
	{
		return new vec2(this.x / vec.x, this.y / vec.y);
	}

	add(vec)
	{
		return new vec2(this.x + vec.x, this.y + vec.y);
	}

	subtract(vec)
	{
		return new vec2(this.x - vec.x, this.y - vec.y);
	}
	
	clamp(high, low)
	{
		return new vec2(
			Math.min(Math.max(this.x, low.x), high.x),
			Math.min(Math.max(this.y, low.y), high.y)
		)
	}

	length()
	{
		return Math.sqrt(this.x * this.x + this.y * this.y);
	}

	distTo(vec)
	{
		return this.subtract(vec).length();
	}
}
